### PR TITLE
Add MIR Validate statement

### PR DESCRIPTION
**Pull request by RalfJung**
This adds statements to MIR that express when types are to be validated (following [Types as Contracts](https://internals.rust-lang.org/t/types-as-contracts/5562)). Obviously nothing is stabilized, and in fact a `-Z` flag has to be passed for behavior to even change at all.

This is meant to make experimentation with Types as Contracts in miri possible. The design is definitely not final.

Cc @nikomatsakis @aturon

### CODE REVIEW

**Comment by oli-obk** [comment]
Tidy fails due to lines longer than 100 chars

**Comment by RalfJung** [comment]
Yeah I find it very hard to follow that requirement. Always feel like I have to make my code ugly to comply. Seems like I missed some places. :/

**Comment by RalfJung** [comment]
Ideally I'd like to run AddValidation before ElaborateDrops and after EraseRegions. For this to work, I need to move ElaborateDrops up, but <https://github.com/RalfJung/rust/commit/3eb41412280abdfc3af56b1f5eb413f7a744fede> fails badly -- many tests now report "Broken MIR: moving out of lvalue".

@arielb1 I suppose just naively moving the pass around doesn't work.^^ Anything else I should try?

**Comment by RalfJung** [comment]
With some more help from @arielb1 I managed to reorder the passes without breaking everything (well, the test suite still passes). They also advised me to change `move_path::abs_domain` to do erasure, which was not actually necessary. Unfortunately, doing region erasure requires access to the `TyCtxt`, so this commit touches lots of other code as well. For `MovePathLookup::find` where no `TyCtxt` was available, I was not sure whether it should be added to the struct (so all methods just have it) or as a new parameter to the functions that need it -- I now went with a new parameter.

**Comment by RalfJung** [comment]
This relies on #43512, which however is already r+'ed so it should land soon-ish.

**Comment by RalfJung** [comment]
I rebased to resolve the merge conflict.

File: src/librustc_mir/transform/add_validation.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                          tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                          _: MirSource,
+                          mir: &mut Mir<'tcx>) {
+        if !tcx.sess.opts.debugging_opts.mir_emit_validate {
+            return;
+        }
+
+        let local_decls = mir.local_decls.clone(); // FIXME: Find a way to get rid of this clone.
+
+        /// Convert an lvalue to a validation operand.
```
Comment (kennytm) [inline src/librustc_mir/transform/add_validation.rs]:
```
[00:24:45]    Compiling rustc_mir v0.0.0 (file:///checkout/src/librustc_mir)^^^[(B^[[m
[00:24:45] error: doc comment not used by rustdoc
[00:24:45]   --> /checkout/src/librustc_mir/transform/add_validation.rs:94:9
[00:24:45]    |
[00:24:45] 94 |         /// Convert an lvalue to a validation operand.
[00:24:45]    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:24:45]    |
[00:24:45] note: lint level defined here
[00:24:45]   --> /checkout/src/librustc_mir/lib.rs:20:9
[00:24:45]    |
[00:24:45] 20 | #![deny(warnings)]
[00:24:45]    |         ^^^^^^^^
```

Comment (RalfJung) [inline src/librustc_mir/transform/add_validation.rs]: AH, thanks. Should be fixed with the new commit.

File: src/librustc/mir/mod.rs [hunk 2]
```diff
@@ -818,12 +818,16 @@ pub enum StatementKind<'tcx> {
...
     StorageDead(Lvalue<'tcx>),
 
+    /// Execute a piece of inline Assembly.
     InlineAsm {
         asm: Box<InlineAsm>,
         outputs: Vec<Lvalue<'tcx>>,
         inputs: Vec<Operand<'tcx>>
     },
 
+    /// Assert the given lvalues to be valid inhabitants of their type.
```
Comment (nikomatsakis) [inline src/librustc/mir/mod.rs]: I think we could extend this comment. For example, it's worth saying that this is part of an experimental attempt at giving semantics to unsafe code, and these statements are not generated during normal MIR execution. Maybe include a link to your blog post. =)

File: src/librustc/mir/mod.rs [hunk 3]
```diff
@@ -832,13 +836,56 @@ pub enum StatementKind<'tcx> {
     Nop,
 }
 
+#[derive(Copy, Clone, RustcEncodable, RustcDecodable, PartialEq, Eq)]
+pub enum ValidationOp {
+    Acquire,
```
Comment (nikomatsakis) [inline src/librustc/mir/mod.rs]: Can we document at least a bit the semantics of each of these operations in doc comment form?

File: src/librustc_mir/transform/clean_end_regions.rs [hunk 4]
```diff
@@ -54,13 +56,32 @@ impl<'tcx> Visitor<'tcx> for GatherBorrowedRegions {
...
             }
         }
         self.super_rvalue(rvalue, location);
     }
+
+    fn visit_statement(&mut self,
+                       block: BasicBlock,
+                       statement: &Statement<'tcx>,
+                       location: Location) {
+        self.super_statement(block, statement, location);
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/clean_end_regions.rs]: if this just calls super, should we remove?

File: src/librustc_mir/transform/clean_end_regions.rs [hunk 5]
```diff
@@ -54,13 +56,32 @@ impl<'tcx> Visitor<'tcx> for GatherBorrowedRegions {
...
+    }
+
+    fn visit_ty(&mut self, ty: &Ty<'tcx>, _: Lookup) {
+        // Gather regions that occur in types
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/clean_end_regions.rs]: Am I correct that the idea here is that types have been erased, so we are only keeping those regions that appear in types from validation commands? If so, merits a comment.

File: src/test/mir-opt/validate_3.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+struct Test {
+    x: i32
+}
+
+fn foo(_x: &i32) {}
+
+fn main() {
+    let t = Test { x: 0 };
+    let t = &t;
+    foo(&t.x);
```
Comment (nikomatsakis) [inline src/test/mir-opt/validate_3.rs]: These tests seem good, if basic. I wonder if we want to test any of the more subtle scenarios...? Maybe this is good enough for now. These kinds of tests *are* somewhat brittle -- the main thing I want is some way to ensure we are not bitrotting your code when we rebase. To that end, maybe something that tests a case where we have to "reacquire" the result from a call as well would be a good idea?

e.g., `let x = vec.split_at_mut(3);`

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks pretty good to me. I left some minor comments, requests for documentation, that sort of thing.

File: src/librustc_mir/transform/clean_end_regions.rs (hunk 4 above)
Comment (RalfJung) [inline src/librustc_mir/transform/clean_end_regions.rs]: Ah right. I had refactored this and did not realize I made this the default impl.

File: src/librustc_mir/transform/clean_end_regions.rs (hunk 5 above)
Comment (RalfJung) [inline src/librustc_mir/transform/clean_end_regions.rs]: No, we keep around all EndRegions for regions mentioned in any type anywhere. This still only concerns regions that are local to our function (otherwise, there wouldn't be an EndRegion) -- so, this covers all regions in the current function that are potentially interesting.

The way ordering works in the driver, `EraseEndRegions` runs before `AddValidate`, so we cannot just keep those that appear in validation statements.

File: src/test/mir-opt/validate_3.rs (hunk 6 above)
Comment (RalfJung) [inline src/test/mir-opt/validate_3.rs]: Reacquire happens at the EndRegion, so code generation is no different. I left the relevant EndRegion in the tests for this reason.

I also hope that sooner rather than later after this is merged, the miri test suite will start testing these things, and with miri merged in rustc, maybe we can run the miri test suite with this "run cargo test on external projects" infrastructure we got.

**Comment by RalfJung** [comment]
Something is not right with the way I try to check if a function is unsafe... I am seeing errors like this in miri's test suite:
```
error: internal compiler error: src/librustc/hir/map/mod.rs:757: expected item, found method eq in <P as std::cmp::PartialEq>::eq (id=1531)
```
This is triggered by the `expect_item` in my new pass. So it seems this is not the way to visit a function including its declaration.

**Comment by RalfJung** [comment]
Ah, this still fails to handle closures. I have a patch fixing that, however, there still is a problem: If the closure is entirely contained in an unsafe block, we must consider it unsafe (after all, unsafe operations are permitted). However, when visiting the closure's HIR, we do not actually hit an unsafe block.

**Comment by RalfJung** [comment]
Still failing to build libstd with the flag turned on:
```
error: internal compiler error: src/librustc_mir/transform/add_validation.rs:142: Expected function, method or closure, found struct_ctor mem::Discriminant (id=12499)
```
I am at loss here. Struct ctors are functions that have their own MIR...?

**Comment by arielb1** [comment]
> I am at loss here. Struct ctors are functions that have their own MIR...?

Yes they are. The MIR is created by https://github.com/rust-lang/rust/blob/master/src/librustc_mir/shim.rs

**Comment by RalfJung** [comment]
I see. I fixed the code to handle that, but now I get
```
error: internal compiler error: src/librustc_mir/transform/add_validation.rs:146: Expected function, method or closure, found trait method ne in cmp::PartialEq::ne (id=15796)
```
I suppose defaulted trait methods are yet another category that needs separate handling.

**Comment by RalfJung** [comment]
Okay, now I am seeing really strange behavior... I thought by returning `OnlyBodies` to the intravisitor, I would only visit one function at a time, but that does not seem to be the case. In libcore/slice/sort.rs, around line 567, I see a backtrace like so:
```
   8: <rustc_mir::transform::add_validation::fn_contains_unsafe::FindUnsafe<'b, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_fn
   9: rustc::hir::intravisit::walk_expr
  10: rustc::hir::intravisit::walk_block
  11: rustc::hir::intravisit::walk_expr
  12: rustc::hir::intravisit::walk_expr
  13: rustc::hir::intravisit::walk_block
  14: rustc::hir::intravisit::walk_expr
  15: <rustc_mir::transform::add_validation::fn_contains_unsafe::FindUnsafe<'b, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_fn
  16: rustc::hir::intravisit::walk_item
  17: rustc_mir::transform::add_validation::fn_contains_unsafe
  18: <rustc_mir::transform::add_validation::AddValidation as rustc::mir::transform::MirPass>::run_pass
```

I also just realized that closures in an unsafe fn probably can use unsafe features as well, so I will have to extend my scan into the closure environment accordingly.

**Comment by RalfJung** [comment]
Okay, it seems to all work now. This is ready to be merged from my side. I hope I got all this HIR code right.

I would really like to test whether closures are correctly detected to be unsafe, but unfortunately, mir-opt tests are not suited for this: Closures contain the filename in their type and name, which shows up in the lines we would like to test for.

**Comment by RalfJung** [comment]
Ah dang it, I got another ICE in miri's test suite...
```
error: internal compiler error: src/librustc/hir/map/mod.rs:537: couldn't find node id 0 in the AST map
```

**Comment by RalfJung** [comment]
Fixed that as well (thanks @eddyb)

**Review by nikomatsakis — APPROVED** [review]

**Comment by RalfJung** [comment]
Thanks :)

So, what has to happen now to get r+?

**Comment by alexcrichton** [comment]
It looks like this PR [regressed a number of benchmarks 5-7%](https://perf.rust-lang.org/graphs.html?start=2017-08-03T00%3A00%3A00%2B00%3A00&end=2017-08-05T00%3A00%3A00%2B00%3A00&crates=%7B%22list%22%3A%22All%22%2C%22content%22%3Anull%7D&phases=%7B%22list%22%3A%22All%22%2C%22content%22%3Anull%7D&group_by=crate&type=time&yaxis=pct) on perf.rust-lang.org, @RalfJung do you know if this could perhaps cause a performance impact when not enabled?

**Comment by alexcrichton** [comment]
A more clear way to see this regression is [perhaps here](https://perf.rust-lang.org/graphs.html?start=2017-08-03T00%3A00%3A00%2B00%3A00&end=2017-08-05T00%3A00%3A00%2B00%3A00&crates=%7B%22list%22%3A%22List%22%2C%22content%22%3A%5B%22syntex-0.42.2%40000-base%22%5D%7D&phases=%7B%22list%22%3A%22All%22%2C%22content%22%3Anull%7D&group_by=crate&type=time&yaxis=pct)

**Comment by RalfJung** [comment]
Is this compile tune or run time? I guess compile time, but it's not entirely clear to me.

That is strange, the new pass does nothing when not enabled. There is a slight chance that more `EndRegion` are left around briefly because `CleanEndRegions` is a little more cautious, but they will all be gone by the time `EraseRegions` has happened.

@arielb1 could reordering the passes have a bad effect on performance?

**Comment by kennytm** [comment]
@RalfJung Compile time. In particular it seems to be the LLVM pass that takes longer time.

**Comment by alexcrichton** [comment]
Ok I'll open a dedicated issue for this in that case!

https://github.com/rust-lang/rust/issues/43827
